### PR TITLE
add raftlogdir commandline option

### PR DESCRIPTION
**Pull request by melihbirim**
Hello Everyone, this PR contains a cli option to externalize raft log dirs.

Command-line option(_--raftlogdir_) to geth in order to move **quorum-raft-state,** **raft-wal** and **raft-snap** folders due to having large size.

This commit related with https://github.com/ConsenSys/quorum/issues/1002

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -897,6 +902,26 @@ func MakeDataDir(ctx *cli.Context) string {
...
+			return filepath.Join(path, "rinkeby")
+		}
+		if ctx.GlobalBool(GoerliFlag.Name) {
+			return filepath.Join(path, "goerli")
+		}
+		return path
+	}
+	Fatalf("Cannot determine default raft log directory, please set manually(--raftlogdir)")
+	return ""
+}
```
Comment (prd-fox) [inline cmd/utils/flags.go]: This function doesn't look like it's used anywhere. If so, it should be removed.

File: raft/backend.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ type RaftService struct {
 	calcGasLimitFunc func(block *types.Block) uint64
 }
 
-func New(ctx *node.ServiceContext, chainConfig *params.ChainConfig, raftId, raftPort uint16, joinExisting bool, blockTime time.Duration, e *eth.Ethereum, startPeers []*enode.Node, datadir string, useDns bool) (*RaftService, error) {
+func New(ctx *node.ServiceContext, chainConfig *params.ChainConfig, raftId, raftPort uint16, joinExisting bool, blockTime time.Duration, e *eth.Ethereum, startPeers []*enode.Node, datadir, raftLogDir string, useDns bool) (*RaftService, error) {
```
Comment (prd-fox) [inline raft/backend.go]: Since `datadir` is no longer used, it should be removed.

File: raft/handler.go [hunk 3]
```diff
@@ -99,10 +99,10 @@ var errNoLeaderElected = errors.New("no leader is currently elected")
 // Public interface
 //
 
-func NewProtocolManager(raftId uint16, raftPort uint16, blockchain *core.BlockChain, mux *event.TypeMux, bootstrapNodes []*enode.Node, joinExisting bool, datadir string, minter *minter, downloader *downloader.Downloader, useDns bool) (*ProtocolManager, error) {
-	waldir := fmt.Sprintf("%s/raft-wal", datadir)
-	snapdir := fmt.Sprintf("%s/raft-snap", datadir)
-	quorumRaftDbLoc := fmt.Sprintf("%s/quorum-raft-state", datadir)
+func NewProtocolManager(raftId uint16, raftPort uint16, blockchain *core.BlockChain, mux *event.TypeMux, bootstrapNodes []*enode.Node, joinExisting bool, datadir string, raftLogDir string, minter *minter, downloader *downloader.Downloader, useDns bool) (*ProtocolManager, error) {
```
Comment (prd-fox) [inline raft/handler.go]: Since `datadir` is no longer used, it should be removed.

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -144,6 +144,11 @@ var (
 		Usage: "Data directory for the databases and keystore",
 		Value: DirectoryString(node.DefaultDataDir()),
 	}
+	RaftLogDirFlag = DirectoryFlag{
+		Name:  "raftlogdir",
+		Usage: "Raft log directory for the raft-state, raft-snap and raft-wal folders",
+		Value: DirectoryString(node.DefaultDataDir()),
```
Comment (prd-fox) [inline cmd/utils/flags.go:127]: This sets the default directory to be the same location as the default data directory. If the `datadir` changes, this won't update with it. Keeping that assumption the same, particularly for existing nodes, is important.

I added a possible solution in the other comments.

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -1855,6 +1888,7 @@ func RegisterPermissionService(stack *node.Node, useDns bool) {
 func RegisterRaftService(stack *node.Node, ctx *cli.Context, nodeCfg *node.Config, ethChan chan *eth.Ethereum) {
 	blockTimeMillis := ctx.GlobalInt(RaftBlockTimeFlag.Name)
 	datadir := ctx.GlobalString(DataDirFlag.Name)
+	raftLogDir := ctx.GlobalString(RaftLogDirFlag.Name)
```
Comment (prd-fox) [inline cmd/utils/flags.go]: This should read from the `nodeCfg` using the newly added var.

File: cmd/utils/flags.go [hunk 6]
```diff
@@ -1363,6 +1389,13 @@ func setDataDir(ctx *cli.Context, cfg *node.Config) {
 	}
 }
 
+func setRaftLogDir(ctx *cli.Context, cfg *node.Config) {
+	switch {
+	case ctx.GlobalIsSet(RaftLogDirFlag.Name):
+		cfg.RaftLogDir = ctx.GlobalString(RaftLogDirFlag.Name)
+	}
+}
+
```
Comment (prd-fox) [inline cmd/utils/flags.go]: If the raft directory isn't set, then it could fallback to the data directory (which it can pull from the node config).

**Review by prd-fox — CHANGES_REQUESTED** [review]
Using the new flag works nicely. Just some comments around keeping existing functionality & code improvements.

File: cmd/utils/flags.go (hunk 1 above)
Comment (melihbirim) [reply cmd/utils/flags.go]: Yep, this function is not neccessary.

File: raft/handler.go (hunk 3 above)
Comment (melihbirim) [reply raft/handler.go]: Removed.

File: cmd/utils/flags.go (hunk 6 above)
Comment (melihbirim) [reply cmd/utils/flags.go]: a new check added, if the directory not set, it fetches it from datadir option.

**Review by prd-fox — DISMISSED** [review]

**Comment by prd-fox** [comment]
Just needs you to run `go fmt cmd/utils/flags.go` to format the file and pass the linter.

**Review by prd-fox — DISMISSED** [review]

**Comment by ricardolyn** [comment]
@melihbirim can you update the PR with the latest master to solve the conflicts? which we could include in the future

**Comment by melihbirim** [comment]
> @melihbirim can you update the PR with the latest master to solve the conflicts? which we could include in the future

Hello, @ricardolyn I resolved the conflicts and updated my branch. Would you please check the PR. Thanks.

**Comment by ricardolyn** [comment]
@melihbirim some unit tests are failing. also, it needs to be updated with the latest `master` changes. 

We are still interested in merging this work 👍

**Comment by hemreari** [comment]
Hi @ricardolyn, we have updated with the latest `master` branch and all unit tests are OK.

**Review by ricardolyn — APPROVED** [review]
Looks good 👍

**Pull request merged**